cmd/engine/config.go

```diff
@@ -0,0 +1,138 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"net"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/dagger/dagger/network"
+	"github.com/jackpal/gateway"
+	"github.com/moby/buildkit/cmd/buildkitd/config"
+)
+
+// engineDefaultStateDir is the directory that we map to a volume by default.
+const engineDefaultStateDir = "/var/lib/dagger"
+
+// daggerConfigPath is the path containing Dagger-specific configuration, which
+// might be provided by the user.
+const daggerConfigPath = "/etc/dagger"
+
+// cniConfigPath is the path to Dagger's CNI configuration. It will be
+// generated if one isn't provided.
+var cniConfigPath = filepath.Join(daggerConfigPath, "cni.conflist")
+
+// servicesDNSEnvName is the feature flag for enabling the services network
+// stack.
+const servicesDNSEnvName = "_EXPERIMENTAL_DAGGER_SERVICES_DNS"
+
+func setDaggerDefaults(cfg *config.Config) error {
+	if cfg.Root == "" {
+		cfg.Root = engineDefaultStateDir
+	}
+
+	if os.Getenv(servicesDNSEnvName) != "" {
+		// check if CNI config already exists, just so we can respect a
+		// user-provided config
+		if _, err := os.Stat(cniConfigPath); os.IsNotExist(err) {
+			cni, err := cniConfig(network.Name, network.CIDR)
+			if err != nil {
+				return err
+			}
+
+			if err := os.MkdirAll(filepath.Dir(cniConfigPath), 0700); err != nil {
+				return err
+			}
+
+			if err := os.WriteFile(cniConfigPath, cni, 0600); err != nil {
+				return err
+			}
+		}
+
+		setNetworkDefaults(&cfg.Workers.OCI.NetworkConfig)
+
+		// we don't use containerd, but make it match anyway
+		setNetworkDefaults(&cfg.Workers.Containerd.NetworkConfig)
+	}
+
+	return nil
+}
+
+func setNetworkDefaults(cfg *config.NetworkConfig) {
+	if cfg.Mode == "" {
+		cfg.Mode = "cni"
+	}
+	if cfg.CNIConfigPath == "" {
+		cfg.CNIConfigPath = cniConfigPath
+	}
+	if cfg.CNIPoolSize == 0 {
+		cfg.CNIPoolSize = 16
+	}
+}
+
+func cniConfig(name, subnet string) ([]byte, error) {
+	ip, err := gateway.DiscoverInterface()
+	if err != nil {
```

Comment:
extreme nit: it's technically valid to not have a default gateway (e.g. you have disabled internet access), so maybe just default to 1500 and print a warning log if this goes wrong? If there's any complication around that though then doesn't matter, most likely an obscure case (I think?)

Reply:
Done!